core/src/main/java/com/linecorp/armeria/client/retry/RetryConfig.java

```diff
@@ -0,0 +1,205 @@
+     */
+    public static RetryConfigBuilder<RpcResponse> builderForRpc(
+            RetryRuleWithContent<RpcResponse> retryRuleWithContent) {
+        return new RetryConfigBuilder<>(retryRuleWithContent);
+    }
+
+    /**
+     * Returns a new {@link RetryConfigBuilder} with the default values from Flags.
+     * Uses a {@code RetryRule}.
+     */
+    public static RetryConfigBuilder<RpcResponse> builderForRpc(RetryRule retryRule) {
+        return new RetryConfigBuilder<>(retryRule);
+    }
+
+    /**
+     * Returns a new {@link RetryConfigBuilder} with the default values from Flags.
+     * Uses a {@code RetryRuleWithContent<HttpResponse>}.
+     */
+    public static RetryConfigBuilder<HttpResponse> builderForHttp(
+            RetryRuleWithContent<HttpResponse> retryRuleWithContent) {
+        return new RetryConfigBuilder<>(retryRuleWithContent);
+    }
+
+    /**
+     * Returns a new {@link RetryConfigBuilder} with the default values from Flags.
+     * Uses a {@code RetryRule}.
+     */
+    public static RetryConfigBuilder<HttpResponse> builderForHttp(RetryRule retryRule) {
+        return new RetryConfigBuilder<>(retryRule);
+    }
+
+    static <T extends Response> RetryConfigBuilder<T> builder(RetryRule retryRule) {
+        return new RetryConfigBuilder<>(retryRule);
+    }
+
+    static <T extends Response> RetryConfigBuilder<T> builder(
+            RetryRuleWithContent<T> retryRuleWithContent) {
+        return new RetryConfigBuilder<>(retryRuleWithContent);
+    }
+
+    private final int maxTotalAttempts;
+    private final long responseTimeoutMillisForEachAttempt;
+    private final int maxContentLength;
+    private final boolean needsContentInRule;
+
+    @Nullable
+    private final RetryRule retryRule;
+
+    @Nullable
+    private final RetryRuleWithContent<T> retryRuleWithContent;
+
+    @Nullable
+    private final RetryRule fromRetryRuleWithContent;
+
+    RetryConfig(RetryRule retryRule, int maxTotalAttempts, long responseTimeoutMillisForEachAttempt) {
+        checkArguments(maxTotalAttempts, responseTimeoutMillisForEachAttempt);
+        this.retryRule = requireNonNull(retryRule, "retryRule");
+        this.maxTotalAttempts = maxTotalAttempts;
+        this.responseTimeoutMillisForEachAttempt = responseTimeoutMillisForEachAttempt;
+        needsContentInRule = false;
+        maxContentLength = 0;
+        retryRuleWithContent = null;
+        fromRetryRuleWithContent = null;
+    }
+
+    RetryConfig(
+            RetryRuleWithContent<T> retryRuleWithContent,
+            int maxContentLength,
+            int maxTotalAttempts,
+            long responseTimeoutMillisForEachAttempt) {
+        checkArguments(maxTotalAttempts, responseTimeoutMillisForEachAttempt);
+        this.maxContentLength = maxContentLength;
+        this.retryRuleWithContent = requireNonNull(retryRuleWithContent, "retryRuleWithContent");
+        fromRetryRuleWithContent = RetryRuleUtil.fromRetryRuleWithContent(retryRuleWithContent);
```

Comment:
I think we don't need to keep `fromRetryRuleWithContent` as a member field. How about removing `fromRetryRuleWithContent` and making `retryRule` not null always? i.e.,
```java
retryRule = RetryRuleUtil.fromRetryRuleWithContent(retryRuleWithContent);
```

Reply:
I agree this would be simpler. The problem is, while fromRetryRuleWithContent() is internal, retryRule() is a public method, and it is very confusing for it to not return null when a retryRule is not set. So I think we can pay the price of slightly more complex internal implementation for the sake of a clearer API.

Reply:
Yeah. That is a good point. Let's keep the current API